apps/prairielearn/src/ee/pages/lti13CourseNavigation/lti13CourseNavigation.ts

```diff
@@ -0,0 +1,198 @@
+      if (role_instructor) {
+        await queryAsync(sql.upsert_lci, {
+          lti13_instance_id: req.params.lti13_instance_id,
+          course_instance_id: lci.course_instance_id,
+          deployment_id: lti13_claims['https://purl.imsglobal.org/spec/lti/claim/deployment_id'],
+          context_id: lti13_claims['https://purl.imsglobal.org/spec/lti/claim/context'].id,
+          context_label: lti13_claims['https://purl.imsglobal.org/spec/lti/claim/context'].label,
+          context_title: lti13_claims['https://purl.imsglobal.org/spec/lti/claim/context'].title,
+        });
+      }
+
+      // Redirect to linked course instance
+      res.redirect(
+        `/pl/course_instance/${lci.course_instance_id}/${role_instructor ? 'instructor/' : ''}`,
+      );
+      return;
+    }
+
+    let courseName = 'your course';
+    if (
+      'label' in lti13_claims['https://purl.imsglobal.org/spec/lti/claim/context'] &&
+      'title' in lti13_claims['https://purl.imsglobal.org/spec/lti/claim/context']
+    ) {
+      courseName = lti13_claims['https://purl.imsglobal.org/spec/lti/claim/context'].label;
+      courseName += `: ${lti13_claims['https://purl.imsglobal.org/spec/lti/claim/context'].title}`;
+    }
+
+    if (!role_instructor) {
+      // Students get a "come back later" message
+      res.send(
+        Lti13CourseNavigationNotReady({
+          resLocals: res.locals,
+          courseName,
+        }),
+      );
+    } else {
+      let courses = await selectCoursesWithEditAccess({
+        user_id: res.locals.authn_user.user_id,
+        is_administrator: res.locals.authn_is_administrator,
+      });
+
+      // FIXME
+      if ('nocourse' in req.query) {
+        courses = [];
+      }
+
+      // FIXME
+      let course_instances: CourseInstance[] = [];
+
+      for (const course of courses) {
+        const loopCI = await selectCourseInstancesWithStaffAccess({
+          course_id: course.id,
+          user_id: res.locals.authn_user.user_id,
+          authn_user_id: res.locals.authn_user.user_id,
+          is_administrator: res.locals.authn_is_administrator,
+          authn_is_administrator: res.locals.authn_is_administrator,
+        });
+
+        course_instances = [...course_instances, ...loopCI];
+      }
+
+      res.send(
+        Lti13CourseNavigationInstructor({
+          resLocals: res.locals,
+          courseName,
+          courses,
+          course_instances,
+        }),
+      );
+    }
+  }),
+);
+
+router.post(
+  '/',
+  asyncHandler(async (req, res) => {
+    const unsafe_course_instance_id = req.body.ci_id;
+    const unsafe_lti13_instance_id = req.params.lti13_instance_id;
+
+    // FIXME
+    const lti13_claims = devel_lti13_claims;
+    const authn_lti13_instance_id = devel_authn_lti13_instance_id;
+
+    // Validate user login match this lti13_instance
+    if (unsafe_lti13_instance_id !== authn_lti13_instance_id) {
```

Comment:
Doing this security check, I realized I didn't have an easy way after auth with only the LTI metadata to map the user back to their lti13_instance. This was added to #8855, discussed here https://github.com/PrairieLearn/PrairieLearn/pull/8855#discussion_r1431861511 and mocked up in this PR.

Reply:
If both the body and the session are missing this value, we'll have `undefined === undefined`, which is always true. Does that give us the desired behavior here?

Reply:
I'm not sure the best way to fix this so leaving it for more discussion/advice.

Reply:
I'd be inclined to add explicit `== null` checks here for each. GitHub won't let me leave a suggestion here, but essentially:

```ts
    // Validate user login matches this lti13_instance
    if (unsafe_lti13_instance_id == null || authn_lti13_instance_id == null || unsafe_lti13_instance_id !== authn_lti13_instance_id) {
      throw error.make(403, 'Permission denied');
    }
```

You'll need to run Prettier and all that.

Reply:
From another comment, I passed the `authn_lti13_instance_id` into the SQL where we were looking things up and removed the checking in JS. Does that work for you?

Reply:
https://github.com/PrairieLearn/PrairieLearn/pull/9098#discussion_r1626638137